lib/ansible/plugins/connection/lxd.py

```diff
@@ -0,0 +1,116 @@
+# (c) 2016 Matt Clay <matt@mystile.com>
+#
+# This file is part of Ansible
+#
+# Ansible is free software: you can redistribute it and/or modify
+# it under the terms of the GNU General Public License as published by
+# the Free Software Foundation, either version 3 of the License, or
+# (at your option) any later version.
+#
+# Ansible is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with Ansible.  If not, see <http://www.gnu.org/licenses/>.
+
+from __future__ import (absolute_import, division, print_function)
+__metaclass__ = type
+
+import os
+from distutils.spawn import find_executable
+from subprocess import call, Popen, PIPE
+
+import ansible.constants as C
+from ansible.errors import AnsibleError, AnsibleFileNotFound
+from ansible.plugins.connection import ConnectionBase
+from ansible.utils.unicode import to_str, to_bytes, to_unicode
+
+try:
+    from __main__ import display
+except ImportError:
+    from ansible.utils.display import Display
+    display = Display()
+
+
+class Connection(ConnectionBase):
+    """ lxd based connections """
+
+    transport = "lxd"
+    has_pipelining = True
+
+    def __init__(self, play_context, new_stdin, *args, **kwargs):
+        super(Connection, self).__init__(play_context, new_stdin, *args, **kwargs)
+
+        self._host = self._play_context.remote_addr
+        self._lxc_cmd = find_executable("lxc")
+
+        if not self._lxc_cmd:
+            raise AnsibleError("lxc command not found in PATH")
+
+    def _connect(self):
+        """connect to lxd (nothing to do here) """
+        super(Connection, self)._connect()
+
+        self._connected = True
+
+    def exec_command(self, cmd, in_data=None, sudoable=True):
+        """ execute a command on the lxd host """
+        super(Connection, self).exec_command(cmd, in_data=in_data, sudoable=sudoable)
+
+        display.vvv(u"EXEC {0}".format(cmd), host=self._host)
+
+        executable = C.DEFAULT_EXECUTABLE.split()[0] if C.DEFAULT_EXECUTABLE else "/bin/sh"
+
+        local_cmd = [self._lxc_cmd, "exec", self._host, "--", executable, "-c", cmd]
+
+        local_cmd = map(to_bytes, local_cmd)
+        in_data = to_bytes(in_data)
```

Comment:
I've fixed the to_bytes call on in_data to return None if that's what's given. This is the same as what the ssh plugin does.